<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Шаг-1.-Обзор-данных" data-toc-modified-id="Шаг-1.-Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Шаг 1. Обзор данных</a></span><ul class="toc-item"><li><span><a href="#Составим-представление-о-входных-данных." data-toc-modified-id="Составим-представление-о-входных-данных.-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Составим представление о входных данных.</a></span></li><li><span><a href="#Проверим-в-таблице-наличие-пропусков." data-toc-modified-id="Проверим-в-таблице-наличие-пропусков.-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Проверим в таблице наличие пропусков.</a></span></li><li><span><a href="#Проверим-в-таблице-наличие-дубликатов" data-toc-modified-id="Проверим-в-таблице-наличие-дубликатов-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Проверим в таблице наличие дубликатов</a></span></li><li><span><a href="#Проверим-качество-данных-в-каждом-столбце" data-toc-modified-id="Проверим-качество-данных-в-каждом-столбце-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Проверим качество данных в каждом столбце</a></span><ul class="toc-item"><li><span><a href="#children" data-toc-modified-id="children-1.4.1"><span class="toc-item-num">1.4.1&nbsp;&nbsp;</span><code>children</code></a></span></li><li><span><a href="#days_employed" data-toc-modified-id="days_employed-1.4.2"><span class="toc-item-num">1.4.2&nbsp;&nbsp;</span><code>days_employed</code></a></span></li><li><span><a href="#dob_years" data-toc-modified-id="dob_years-1.4.3"><span class="toc-item-num">1.4.3&nbsp;&nbsp;</span><code>dob_years</code></a></span></li><li><span><a href="#education" data-toc-modified-id="education-1.4.4"><span class="toc-item-num">1.4.4&nbsp;&nbsp;</span><code>education</code></a></span></li><li><span><a href="#education_id" data-toc-modified-id="education_id-1.4.5"><span class="toc-item-num">1.4.5&nbsp;&nbsp;</span><code>education_id</code></a></span></li><li><span><a href="#family_status" data-toc-modified-id="family_status-1.4.6"><span class="toc-item-num">1.4.6&nbsp;&nbsp;</span><code>family_status</code></a></span></li><li><span><a href="#family_status_id" data-toc-modified-id="family_status_id-1.4.7"><span class="toc-item-num">1.4.7&nbsp;&nbsp;</span><code>family_status_id</code></a></span></li><li><span><a href="#gender" data-toc-modified-id="gender-1.4.8"><span class="toc-item-num">1.4.8&nbsp;&nbsp;</span><code>gender</code></a></span></li><li><span><a href="#income_type" data-toc-modified-id="income_type-1.4.9"><span class="toc-item-num">1.4.9&nbsp;&nbsp;</span><code>income_type</code></a></span></li><li><span><a href="#debt" data-toc-modified-id="debt-1.4.10"><span class="toc-item-num">1.4.10&nbsp;&nbsp;</span><code>debt</code></a></span></li><li><span><a href="#total_income" data-toc-modified-id="total_income-1.4.11"><span class="toc-item-num">1.4.11&nbsp;&nbsp;</span><code>total_income</code></a></span></li><li><span><a href="#purpose" data-toc-modified-id="purpose-1.4.12"><span class="toc-item-num">1.4.12&nbsp;&nbsp;</span><code>purpose</code></a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Шаг-2.1-Заполнение-пропусков" data-toc-modified-id="Шаг-2.1-Заполнение-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Шаг 2.1 Заполнение пропусков</a></span></li><li><span><a href="#Шаг-2.2-Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Шаг-2.2-Проверка-данных-на-аномалии-и-исправления.-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Шаг 2.2 Проверка данных на аномалии и исправления.</a></span><ul class="toc-item"><li><span><a href="#gender" data-toc-modified-id="gender-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span><code>gender</code></a></span></li><li><span><a href="#education" data-toc-modified-id="education-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span><code>education</code></a></span></li><li><span><a href="#dob_years" data-toc-modified-id="dob_years-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span><code>dob_years</code></a></span></li><li><span><a href="#children" data-toc-modified-id="children-2.2.4"><span class="toc-item-num">2.2.4&nbsp;&nbsp;</span><code>children</code></a></span></li><li><span><a href="#days_employed" data-toc-modified-id="days_employed-2.2.5"><span class="toc-item-num">2.2.5&nbsp;&nbsp;</span><code>days_employed</code></a></span></li></ul></li><li><span><a href="#Шаг-2.3.-Изменение-типов-данных." data-toc-modified-id="Шаг-2.3.-Изменение-типов-данных.-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Шаг 2.3. Изменение типов данных.</a></span></li><li><span><a href="#Шаг-2.4.-Удаление-дубликатов." data-toc-modified-id="Шаг-2.4.-Удаление-дубликатов.-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Шаг 2.4. Удаление дубликатов.</a></span></li><li><span><a href="#Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Шаг-2.6.-Категоризация-дохода." data-toc-modified-id="Шаг-2.6.-Категоризация-дохода.-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Шаг 2.6. Категоризация дохода.</a></span></li><li><span><a href="#Шаг-2.7.-Категоризация-целей-кредита." data-toc-modified-id="Шаг-2.7.-Категоризация-целей-кредита.-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Шаг 2.7. Категоризация целей кредита.</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Вопрос-1:" data-toc-modified-id="Вопрос-1:-2.8.0.1"><span class="toc-item-num">2.8.0.1&nbsp;&nbsp;</span>Вопрос 1:</a></span></li><li><span><a href="#Вывод-1:" data-toc-modified-id="Вывод-1:-2.8.0.2"><span class="toc-item-num">2.8.0.2&nbsp;&nbsp;</span>Вывод 1:</a></span></li><li><span><a href="#Вопрос-2:" data-toc-modified-id="Вопрос-2:-2.8.0.3"><span class="toc-item-num">2.8.0.3&nbsp;&nbsp;</span>Вопрос 2:</a></span></li><li><span><a href="#Вывод-2:" data-toc-modified-id="Вывод-2:-2.8.0.4"><span class="toc-item-num">2.8.0.4&nbsp;&nbsp;</span>Вывод 2:</a></span></li><li><span><a href="#Вопрос-3:" data-toc-modified-id="Вопрос-3:-2.8.0.5"><span class="toc-item-num">2.8.0.5&nbsp;&nbsp;</span>Вопрос 3:</a></span></li><li><span><a href="#Вывод-3:" data-toc-modified-id="Вывод-3:-2.8.0.6"><span class="toc-item-num">2.8.0.6&nbsp;&nbsp;</span>Вывод 3:</a></span></li><li><span><a href="#Вопрос-4:" data-toc-modified-id="Вопрос-4:-2.8.0.7"><span class="toc-item-num">2.8.0.7&nbsp;&nbsp;</span>Вопрос 4:</a></span></li><li><span><a href="#Вывод-4:" data-toc-modified-id="Вывод-4:-2.8.0.8"><span class="toc-item-num">2.8.0.8&nbsp;&nbsp;</span>Вывод 4:</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

# Исследование надежности заемщиков.

Платежная дисциплина клиентов зависит от можества факторов:
 * доход клиента
 * возраст
 * место работы
 * регион проживания
 * наличие иных обязательств

**Цель исследования**    

Проверить, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.

**Ход исследования**

Входные данные от банка — статистика о платёжеспособности клиентов.<br>
Путь к файлу:  `data.csv`

О качестве данных ничего не известно. Поэтому перед проверкой гипотез будет произведен обзор данных и исправление ошибок (при необходимости).

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Шаг 1. Обзор данных

### Составим представление о входных данных.

Импортируем библиотеку `pandas` и прочитаем файл `data.csv`

In [6]:
# импортируем библиотеку pandas 
import pandas as pd

In [7]:
# записываем данные в переменную df
df = pd.read_csv('data.csv')

Выведем на экран первые и последние 5 строк таблицы:

In [8]:
# получение первых 5 строк таблицы df
df.head()

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [9]:
# получение последних 5 строк таблицы df
df.tail()

children  days_employed  dob_years education  education_id  \
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля

**Описание данных**

**`children`** — количество детей в семье<br>
**`days_employed`** — общий трудовой стаж в днях<br>
**`dob_years`** — возраст клиента в годах<br>
**`education`** — уровень образования клиента<br>
**`education_id`** — идентификатор уровня образования<br>
**`family_status`** — семейное положение<br>
**`family_status_id`** — идентификатор семейного положения<br>
**`gender`** — пол клиента<br>
**`income_type`** — тип занятости<br>
**`debt`** — имел ли задолженность по возврату кредитов<br>
**`total_income`** — ежемесячный доход<br>
**`purpose`** — цель получения кредита<br>

**Вывод:** <br>
Наименование столбцов корректное. <br>
Сразу бросается в глаза то, что в столбце `days_employed` содержаться отрицательные значения.

Получим общую информацию о таблице.

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Вывод:**<br>
В двух столбцах `days_employed` и `total_income`содержаться пропуски.<br>
Это мы еще раз в явном виде проверим на следующем шаге.<br>
После исправления ошибок в этих столбцах приведем их к целочисленному формату `int`.<br>

### Проверим в таблице наличие пропусков.

In [11]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

### Проверим в таблице наличие дубликатов

In [12]:
df.duplicated().sum()

54

**Вывод:**<br>
В таблице задублированно 54 строки.

### Проверим качество данных в каждом столбце

#### `children`

In [13]:
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

**Вывод:**<br>
В столбце `children` содержаться аномальные значения 20 и -1

#### `days_employed`

Отсортируем столбец по возрастанию и выведем первые и последние 10 строк

In [14]:
df['days_employed'].sort_values()[:10]

16335   -18388.949901
4299    -17615.563266
7329    -16593.472817
17838   -16264.699501
16825   -16119.687737
3974    -15835.725775
1539    -15785.678893
4321    -15773.061335
7731    -15618.063786
15675   -15410.040779
Name: days_employed, dtype: float64

In [16]:
df.loc[df['days_employed'] > 0,'days_employed'].sort_values()[-10:]

8369     401590.452231
10991    401591.828457
17823    401614.475622
13420    401619.633298
4697     401635.032697
7794     401663.850046
2156     401674.466633
7664     401675.093434
10006    401715.811749
6954     401755.400475
Name: days_employed, dtype: float64

**Вывод:** <br>
* в столбце содержатся отрицательные значения
* в столбце содержатся аномальные значения: 401755 дня это 1100 лет
* в столбце содержатся пропуски

#### `dob_years`

In [17]:
df['dob_years'].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

**Вывод:** <br>
В столбце содержится значение 0

#### `education`

In [18]:
df['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

**Вывод:**<br>
В столбце содержаться неявные дубликаты из-за разного регистра значений

#### `education_id`

In [19]:
df['education_id'].value_counts()

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

**Вывод:** <br>
Каких-то аномалий не выявлено

#### `family_status`

In [20]:
df['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

**Вывод:** <br>
Значения не в одном регистре но на качество данных это не влияет

#### `family_status_id`

In [21]:
df['family_status_id'].value_counts()

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

**Вывод:**<br>
Аномалий нет

#### `gender`

In [22]:
df['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

**Вывод:**<br>
В столбце содержится один не явный пропуск

#### `income_type`

In [23]:
df['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
в декрете              1
студент                1
Name: income_type, dtype: int64

**Вывод:**<br>
Ошибок не выявлено

#### `debt`

In [24]:
df['debt'].value_counts()

0    19784
1     1741
Name: debt, dtype: int64

**Вывод:**<br>
Ошибок не выявлено

#### `total_income`

In [25]:
total_income_sorted = sorted(df['total_income'])

In [26]:
total_income_sorted[:10]

[34355.21078469274,
 36649.0784390092,
 40940.19996410648,
 45452.471394878914,
 47497.244727000965,
 47766.39152492994,
 48734.45879885573,
 50759.500964816885,
 52461.006385709385,
 55112.757731719255]

In [27]:
total_income_sorted[-10:]

[153864.65032820884,
 155999.80651240775,
 178059.55349137326,
 224791.86238177711,
 236937.2212823449,
 237328.34604625436,
 244093.0505004255,
 322807.77660314646,
 355988.40718762303,
 446870.82368040615]

**Вывод:**<br>
Ошибок не выявлено

#### `purpose`

In [28]:
df['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   653
операции с коммерческой недвижимостью     651
жилье                                     647
покупка жилья                             647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
покупка своего жилья                      620
строительство недвижимости                620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

**Вывод:**<br>
В столбце содержится больше количество неявных дубликатов, но т.к. это поле не понадобится для исследования, то значения в нем исправлять не будем

 ### Вывод

Входные данные содержат необходимую информацию для исследования, но в самих данных есть ошибки. <br>

* В таблице задублированно 54 строки.<br>

* `children` в столбце содержаться аномальные значения 20 и -1

* `days_employed`<br>
    ** в столбце содержаться отрицательные значения<br>
    ** в столбце содержаться аномальные значения: 401755 дня это 1100 лет лет<br>
    ** в столбце содержаться пропуски<br>
    ** формат столбца необходимо привести к целочисленному<br>
* `dob_years` в столбце содержится нулевое значение<br>
* `education` в столбце содержаться неявные дубликаты из-за разного регистра значений<br>
* `gender` в столбце содержится один не явный пропуск (`XNA`)
* `total_income` в столбце содержаться пропущенные значения, формат необходимо привести к целочисленному


##  Предобработка данных

### Шаг 2.1 Заполнение пропусков

Заполним пропуски в столбце `total_income`. Доля пропусков составляет 10%, поэтому просто удалить строки нельзя. В отсортированных данных этого столбца мы видели, что есть небольшое количество значаний, которые значительно больше всех остальных (больше 200 000), поэтому среднее будет несовсем корректно применять, т.к. из-за экстримальных значений оно будет искажаться в большую сторону. Заполним пропуски медианными значениеми по каждому типу дохода, у которого есть пропуски по размеру дохода. Ниже обратим внимание, что доход не заполнен для тех строк, в которых не указан стаж работы.

In [30]:
df['total_income'].isna().sum()

2174

In [31]:
# определим типы доходов, для которых встречаеются пропуски размера дохода
# затем для каждой такой категории посчитаем медианное значение
# и в конце для каждого типа дохода, для которого пропущен размер дохода, заменим соответсвующим
# медианным значением для такого типа дохода
for i in df.loc[df['total_income'].isna(),'income_type'].unique():
    median_by_type = df.loc[df['income_type'] == i,'total_income'].median()
    df.loc[df['income_type'] == i,'total_income'] = df.loc[df['income_type'] == i,'total_income'].fillna(median_by_type)
    

Убедимся, что в поле `total_income` не осталось пропусков

In [32]:
df.loc[df['total_income'].isna()]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

### Шаг 2.2 Проверка данных на аномалии и исправления.

#### `gender`

В выводах к Шагу 1 мы отметили, что в поле `gener` содержится значение `XNA`.
Удалим эту строку.

In [36]:
df = df[df['gender'] != 'XNA']

#### `education`

Приведем все значения в этом поле к одному регистру (нижнему).

In [37]:
df['education'] = df['education'].str.lower()

Проверим результат исправления

In [38]:
df['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      743
начальное                282
ученая степень             6
Name: education, dtype: int64

#### `dob_years`

На Шаге 1 мы установили, что для 101 записи возраст заполнен 0. Это меньше 1% от общего размера данных. Можно было бы пренебречь этими записям, но чтобы не терять записи для анализа поставленной задачи, заменим 0 на среднее значение возвраста в разрезе типов дохода. 

In [39]:
for i in df.loc[df['dob_years'] == 0,'income_type'].unique():
    mean_by_type = df.loc[df['income_type'] == i,'dob_years'].mean()
    df.loc[(df['income_type'] == i) & (df['dob_years'] == 0),'dob_years'] = mean_by_type

Проверим, что нулевых значений не осталось.

In [40]:
df.loc[df['dob_years'] == 0]['dob_years'].count()

0

#### `children`

Заменим значение -1 на 1

In [41]:
df.loc[df['children'] == -1, 'children'] = 1

Заменим 20 на среднее значение для каждой категории `family_status`

In [42]:
for i in df.loc[df['children'] == 20,'family_status'].unique():
    mean_by_status = df.loc[(df['family_status'] == i) & ~(df['children'] == 20),'children'].mean()
    df.loc[(df['family_status'] == i) & (df['children'] == 20),'children'] = mean_by_status

Проверим результат

In [43]:
df.loc[df['children'] == 20] ['children'].count()

0

#### `days_employed`

Заменим отрицательные значения на положительные

In [44]:
df['days_employed'] = df['days_employed'].abs()

Стаж в 50 лет составляет 18250 дней. Отберем значения стажа выше 20000, отсортируем и посмотрим начальные значения.

In [46]:
df.loc[df['days_employed'] > 20000,'days_employed'].sort_values()[:10]

20444    328728.720605
9328     328734.923996
17782    328771.341387
14783    328795.726728
7229     328827.345667
5332     328834.274455
5756     328842.364916
13325    328862.571412
3642     328874.270668
3386     328895.862752
Name: days_employed, dtype: float64

Можно предположить, что для записей со стажем свыше 20000 дней стаж указан не в днях а в часах. Разделим эти значения на 24.

In [47]:
df.loc[df['days_employed'] > 20000,'days_employed'] = df.loc[df['days_employed'] > 20000,'days_employed'] / 24

In [48]:
df.loc[df['days_employed'] > 0,'days_employed'].sort_values()[-10:]

17823    16733.936484
13420    16734.151387
4697     16734.793029
7794     16735.993752
2156     16736.436110
7664     16736.462226
10006    16738.158823
6954     16739.808353
4299     17615.563266
16335    18388.949901
Name: days_employed, dtype: float64

Заменим пропуски медианным значением

In [49]:
df['days_employed'] = df['days_employed'].fillna(df['days_employed'].median())

Проверим, что пропусков не осталось

In [45]:
df['days_employed'].isna().sum()

0

### Шаг 2.3. Изменение типов данных.

Еще раз проверим тип данных

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21524 non-null  float64
 1   days_employed     21524 non-null  float64
 2   dob_years         21524 non-null  float64
 3   education         21524 non-null  object 
 4   education_id      21524 non-null  int64  
 5   family_status     21524 non-null  object 
 6   family_status_id  21524 non-null  int64  
 7   gender            21524 non-null  object 
 8   income_type       21524 non-null  object 
 9   debt              21524 non-null  int64  
 10  total_income      21524 non-null  float64
 11  purpose           21524 non-null  object 
dtypes: float64(4), int64(3), object(5)
memory usage: 2.1+ MB


Переведем в целочисленный формат значения в столбцах `days_employed`, `children`, `dob_years` и `total_income`

In [51]:
df['days_employed'] = df['days_employed'].astype('int')
df['dob_years'] = df['dob_years'].astype('int')
df['total_income'] =df['total_income'].astype('int')
df['children'] = df['children'].astype('int')

Проверим, что тип данных изменился

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21524 non-null  int32 
 1   days_employed     21524 non-null  int32 
 2   dob_years         21524 non-null  int32 
 3   education         21524 non-null  object
 4   education_id      21524 non-null  int64 
 5   family_status     21524 non-null  object
 6   family_status_id  21524 non-null  int64 
 7   gender            21524 non-null  object
 8   income_type       21524 non-null  object
 9   debt              21524 non-null  int64 
 10  total_income      21524 non-null  int32 
 11  purpose           21524 non-null  object
dtypes: int32(4), int64(3), object(5)
memory usage: 1.8+ MB


### Шаг 2.4. Удаление дубликатов.

Удалим задублированные строки и обновим индексацию

In [53]:
df = df.drop_duplicates().reset_index(drop=True)

Проверим, что дубликатов не осталось

In [54]:
df.duplicated().sum()

0

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Создадим два новых датафрейма, в которых:
каждому уникальному значению из `education` соответствует уникальное значение `education_id` — в первом;
каждому уникальному значению из `family_status` соответствует уникальное значение `family_status_id` — во втором.

In [55]:
df_education = df[['education','education_id']]
df_family_status = df[['family_status','family_status_id']]
df_education = df_education.drop_duplicates().reset_index(drop=True)
df_family_status = df_family_status.drop_duplicates().reset_index(drop=True)

Проверим результа

In [56]:
df_education

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [57]:
df_family_status

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4

Удалим `education` и `family_status` из `df`

In [58]:
df.drop(columns=['education', 'family_status'])

children  days_employed  dob_years  education_id  family_status_id  \
0             1           8437         42             0                 0   
1             1           4024         36             1                 0   
2             0           5623         33             1                 0   
3             3           4124         32             1                 0   
4             0          14177         53             1                 1   
...         ...            ...        ...           ...               ...   
21448         1           4529         43             1                 1   
21449         0          14330         67             1                 0   
21450         1           2113         38             1                 1   
21451         3           3112         38             1                 0   
21452         2           1984         40             1                 0   

      gender income_type  debt  total_income                       purpose  
0          F   сотрудник     0        253875                 покупка жилья  
1          F   сотрудник     0        112080       приобретение автомобиля  
2          M   сотрудник     0        145885                 покупка жилья  
3          M   сотрудник     0        267628    дополнительное образование  
4          F   пенсионер     0        158616               сыграть свадьбу  
...      ...         ...   ...           ...                           ...  
21448      F   компаньон     0        224791             операции с жильем  
21449      F   пенсионер     0        155999          сделка с автомобилем  
21450      M   сотрудник     1         89672                  недвижимость  
21451      M   сотрудник     1        244093  на покупку своего автомобиля  
21452      F   сотрудник     0         82047         на покупку автомобиля  

[21453 rows x 10 columns]

### Шаг 2.6. Категоризация дохода.

На основании диапазонов, указанных ниже, создадим столбец total_income_category с категориями:<br>
0–30000 — 'E';<br>
30001–50000 — 'D';<br>
50001–200000 — 'C';<br>
200001–1000000 — 'B';<br>
1000001 и выше — 'A'.<br>

In [59]:
def total_income_group(total_income):
    if total_income <= 30000:
        return 'E'
    if total_income <= 50000:
        return 'D'
    if total_income <= 200000:
        return 'C'
    if total_income <= 1000000:
        return 'B'
    return 'A'

df['total_income_category'] = df['total_income'].apply(total_income_group)

In [60]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0          14177         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category  
0               покупка жилья                     B  
1     приобретение автомобиля                     C  
2               покупка жилья                     C  
3  дополнительное образование                     B  
4             сыграть свадьбу                     C

### Шаг 2.7. Категоризация целей кредита.

Создадим функцию, которая на основании данных из столбца purpose сформирует новый столбец purpose_category, в который войдут следующие категории: <br>
'операции с автомобилем',<br>
'операции с недвижимостью',<br>
'проведение свадьбы',<br>
'получение образования'.<br>

In [61]:
def purpose_group(purpose):
    if purpose.find('автомобил') >= 0:
        return 'операции с автомобилем'
    if purpose.find('недвижим') >= 0 or purpose.find('жиль') >= 0:
        return 'операции с недвижимостью'
    if purpose.find('свадьб') >= 0:
        return 'проведение свадьбы'
    if purpose.find('образован') >= 0:
        return 'получение образования'
    
df['purpose_category'] = df['purpose'].apply(purpose_group)   

Проверим

In [62]:
df_purpose_group = df[['purpose','purpose_category']]
df_purpose_group = df_purpose_group.drop_duplicates().reset_index(drop=True)
df_purpose_group

purpose          purpose_category
0                            покупка жилья  операции с недвижимостью
1                  приобретение автомобиля    операции с автомобилем
2               дополнительное образование     получение образования
3                          сыграть свадьбу        проведение свадьбы
4                        операции с жильем  операции с недвижимостью
5                              образование     получение образования
6                    на проведение свадьбы        проведение свадьбы
7                  покупка жилья для семьи  операции с недвижимостью
8                     покупка недвижимости  операции с недвижимостью
9        покупка коммерческой недвижимости  операции с недвижимостью
10              покупка жилой недвижимости  операции с недвижимостью
11  строительство собственной недвижимости  операции с недвижимостью
12                            недвижимость  операции с недвижимостью
13              строительство недвижимости  операции с недвижимостью
14      на покупку подержанного автомобиля    операции с автомобилем
15            на покупку своего автомобиля    операции с автомобилем
16   операции с коммерческой недвижимостью  операции с недвижимостью
17        строительство жилой недвижимости  операции с недвижимостью
18                                   жилье  операции с недвижимостью
19         операции со своей недвижимостью  операции с недвижимостью
20                              автомобили    операции с автомобилем
21                   заняться образованием     получение образования
22        сделка с подержанным автомобилем    операции с автомобилем
23                   получение образования     получение образования
24                              автомобиль    операции с автомобилем
25                                 свадьба        проведение свадьбы
26   получение дополнительного образования     получение образования
27                    покупка своего жилья  операции с недвижимостью
28                операции с недвижимостью  операции с недвижимостью
29           получение высшего образования     получение образования
30                         свой автомобиль    операции с автомобилем
31                    сделка с автомобилем    операции с автомобилем
32                  профильное образование     получение образования
33                      высшее образование     получение образования
34                 покупка жилья для сдачи  операции с недвижимостью
35                   на покупку автомобиля    операции с автомобилем
36                            ремонт жилью  операции с недвижимостью
37            заняться высшим образованием     получение образования

### Ответы на вопросы.

##### Вопрос 1:

Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [63]:
data_grouped_children = df.groupby(['children']).agg({'debt':['sum','count']})
data_grouped_children['debt_rate'] = 100 * data_grouped_children['debt']['sum'] / data_grouped_children['debt']['count']
data_grouped_children

debt        debt_rate
           sum  count          
children                       
0         1071  14166  7.560356
1          445   4855  9.165808
2          194   2052  9.454191
3           27    330  8.181818
4            4     41  9.756098
5            0      9  0.000000

Из таблицы мы видим, что клиенты у которых есть 1,2, или 4 ребенка - реже остальных возращают кредиты.<br>
Клиенты, у которых 3 ребенка - возвращают кредит реже, чем клиенты без детей.<br>
Клиенты с 5 детьми кредиты возвращают чаще остальных.

##### Вывод 1:

Клиенты, у которых до 4х детей, возвращают кредиты реже, чем клиенты без детей.

##### Вопрос 2:

Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [64]:
data_grouped_fs = df.groupby(['family_status']).agg({'debt':['sum','count']})
data_grouped_fs['debt_rate'] = 100 * data_grouped_fs['debt']['sum'] / data_grouped_fs['debt']['count']
data_grouped_fs

debt        debt_rate
                       sum  count          
family_status                              
Не женат / не замужем  274   2810  9.750890
в разводе               85   1195  7.112971
вдовец / вдова          63    959  6.569343
гражданский брак       388   4150  9.349398
женат / замужем        931  12339  7.545182

##### Вывод 2:

Вдовы возвращают кредиты чаще других.
А не женатые, не замужние или в гражданском браке - реже остальных

##### Вопрос 3:

Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [65]:
data_grouped_income = df.groupby(['total_income_category']).agg({'debt':['sum','count']})
data_grouped_income['debt_rate'] = 100 * data_grouped_income['debt']['sum'] / data_grouped_income['debt']['count']
data_grouped_income

debt        debt_rate
                        sum  count          
total_income_category                       
A                         2     25  8.000000
B                       356   5041  7.062091
C                      1360  16015  8.492039
D                        21    350  6.000000
E                         2     22  9.090909

##### Вывод 3:

Клиенты с минимальлным уровнем дохода возращают кредиты хуже устальных. Но при это с уровнем дохода выше минимального - возвращают кредиты чаще остальных. При это прямой зависимости между ростом дохода и платежной дисциплиной не наблюдается

##### Вопрос 4:

Как разные цели кредита влияют на его возврат в срок?

In [66]:
data_grouped_purpose = df.groupby(['purpose_category']).agg({'debt':['sum','count']})
data_grouped_purpose['debt_rate'] = 100 * data_grouped_purpose['debt']['sum'] / data_grouped_purpose['debt']['count']
data_grouped_purpose

debt        debt_rate
                          sum  count          
purpose_category                              
операции с автомобилем    403   4306  9.359034
операции с недвижимостью  782  10810  7.234043
получение образования     370   4013  9.220035
проведение свадьбы        186   2324  8.003442

##### Вывод 4:

Кредиты на проведение свадьбы и операций с недвижемостью чаще возвращаются в срок чем на кредиты на иные цели.

## Общий вывод:

In [67]:
data_grouped_fs_chldr = df.groupby([ 'children','family_status']).agg({'debt':['sum','count']})
data_grouped_fs_chldr['debt_rate'] = 100 * data_grouped_fs_chldr['debt']['sum'] / data_grouped_fs_chldr['debt']['count']
data_pivot = data_grouped_fs_chldr.pivot_table(index=['family_status'], columns='children', values='debt_rate')
data_pivot

\
children                      0          1          2          3          4   
family_status                                                                 
Не женат / не замужем  9.291061  11.453744  12.000000  12.500000  50.000000   
в разводе              7.124682   6.645570   8.641975   9.090909   0.000000   
вдовец / вдова         6.227967   8.641975  15.000000   0.000000   0.000000   
гражданский брак       8.464064  11.800000   8.746356  14.285714   0.000000   
женат / замужем        6.904350   8.222370   9.458578   6.827309  10.344828   

                            
children                 5  
family_status               
Не женат / не замужем  NaN  
в разводе              NaN  
вдовец / вдова         NaN  
гражданский брак       0.0  
женат / замужем        0.0

В целом чем больше детей, тем больше вероятность невозврата. <br>

Но для определенных групп возврат 100%:<br>
* в разводе и 4 ребенка
* овдовевшие и 3-4 ребенка
* в гражданском браке 4-5 ребенка
* в браке и 5 детей

Самый большой процент невозврата у клиентов, не состощих в отношениях и 4 детьми.
Люди состоящие в браке или ранее состоявшие чаще возвращают кредит своевременно, чем  те, кто не в отношениях или в гражданском браке.